<a href="https://colab.research.google.com/github/ProgrammerDavid1/Data-Analysis/blob/main/Data%20Analyze%203)%20Data%20Preprocessing%20.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 데이터분석프로그래밍 7주차 실습

## 실습 시작 전 유의사항

※ 반드시 상단 메뉴에서 '**파일 - 드라이브에 사본 저장**' 진행 후 시작하세요.

## 실습 개요
- 데이터 전처리
  - 누락 데이터 처리
  - 범주형 데이터 처리
  - 데이터 정규화

---

## 데이터 전처리

seaborn에서 제공하는 planets 데이터셋을 바탕으로 실습을 진행해볼 것입니다.

아래 코드와 같이 **`load_dataset`** 함수를 통해 seaborn에서 제공하는 여러 데이터셋을 불러올 수 있습니다.

https://github.com/mwaskom/seaborn-data

해당 링크에서 불러올 수 있는 데이터셋들을 확인할 수 있습니다.

In [ ]:
# 데이터셋 준비
import pandas as pd
import seaborn as sns
df = sns.load_dataset('planets')

df.head(5)

method  number  orbital_period   mass  distance  year
0  Radial Velocity       1         269.300   7.10     77.40  2006
1  Radial Velocity       1         874.774   2.21     56.95  2008
2  Radial Velocity       1         763.000   2.60     19.84  2011
3  Radial Velocity       1         326.030  19.40    110.62  2007
4  Radial Velocity       1         516.220  10.50    119.47  2009

### 누락 데이터 처리

데이터셋은 여러 이유로 종종 누락된 데이터가 존재할 수 있습니다.

누락 데이터를 처리하지 않으면 분석을 방해하거나 데이터의 품질을 저하하는 등의 문제를 일으킵니다.

이번 실습에서는 이런 결측치를 처리하는 대표적인 두 가지 방법을 소개하겠습니다.

### 1. 누락 데이터가 포함된 행 제거하기

  - Step 1) DataFrame의 **`isnull`** 함수와 **`sum`** 함수를 사용하여 누락된 데이터의 수를 확인하고,
  - Step 2) **`dropna`** 함수를 통해 누락된 열 중 **가장 수가 적은 두 column**들에 대하여 **둘 중 하나라도 누락된** 행들을 제거 해보세요.

  ex) 열 별 누락된 개수가 `number: 100, mass: 6, year: 30`이라면
  mass와 year 열에 대해, 둘 중 하나라도 누락된 데이터를 찾아 제거

**Step 1 Hint)**
- **`isnull`** 함수는, 각 데이터가 NaN인지 확인해서 True 또는 False로 나타내줍니다.
- **`sum`** 함수는, axis 매개변수를 통해 열 또는 행별로 합계를 구해줍니다.
- 수로 나타내면 True는 1이고, False는 0이 되는 것을 통해 데이터의 총 NaN 개수를 알 수 있습니다.

In [ ]:
# Step 1. isnull 함수와 sum 함수를 통해 누락 데이터 확인
print("열별 누락된 데이터 개수")
print("=======================")
print(df.isnull().sum(axis=0))
print("=======================")

열별 누락된 데이터 개수
method              0
number              0
orbital_period     43
mass              522
distance          227
year                0
dtype: int64


**Step 2 Hint)**
- dropna 함수는 누락된 데이터를 제거할 때 사용되는 함수입니다.
- 매개변수 subset은 어떤 열을 기준으로 결측치를 확인할 지 정해줍니다.\
  여러 열을 기준으로 하려면 리스트로 넣을 수 있습니다.
- 매개변수 how가 "any"면, 기준 중 아무 하나를 만족하면 제거합니다.\
  '"all"이면, 기준 모두를 만족해야 제거합니다.

In [ ]:
# Step 2. dropna 함수를 통해 특정 열에 대해서만 누락 데이터 제거
dropped_df = df.dropna(subset = ['orbital_period', 'distance'], how = 'any', axis = 0)

print(dropped_df)
# 출력의 마지막이 [776 rows x 6 columns] 가 나와야 정답입니다.

               method  number  orbital_period   mass  distance  year
0     Radial Velocity       1      269.300000   7.10     77.40  2006
1     Radial Velocity       1      874.774000   2.21     56.95  2008
2     Radial Velocity       1      763.000000   2.60     19.84  2011
3     Radial Velocity       1      326.030000  19.40    110.62  2007
4     Radial Velocity       1      516.220000  10.50    119.47  2009
...               ...     ...             ...    ...       ...   ...
1030          Transit       1        3.941507    NaN    172.00  2006
1031          Transit       1        2.615864    NaN    148.00  2007
1032          Transit       1        3.191524    NaN    174.00  2007
1033          Transit       1        4.125083    NaN    293.00  2008
1034          Transit       1        4.187757    NaN    260.00  2008

[776 rows x 6 columns]


이런 방법 외에도 DataFrame의 **`info`** 함수를 통해, 각 열에 Non-Null 데이터가 몇 개인지 알 수 있습니다.

이를 바탕으로 현재 **mass** 열에도 아직 Null 데이터가 존재함을 알 수 있습니다.

In [ ]:
dropped_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 776 entries, 0 to 1034
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   method          776 non-null    object 
 1   number          776 non-null    int64  
 2   orbital_period  776 non-null    float64
 3   mass            498 non-null    float64
 4   distance        776 non-null    float64
 5   year            776 non-null    int64  
dtypes: float64(3), int64(2), object(1)
memory usage: 42.4+ KB


In [ ]:
dropped_df.tail(5)

method  number  orbital_period  mass  distance  year
1030  Transit       1        3.941507   NaN     172.0  2006
1031  Transit       1        2.615864   NaN     148.0  2007
1032  Transit       1        3.191524   NaN     174.0  2007
1033  Transit       1        4.125083   NaN     293.0  2008
1034  Transit       1        4.187757   NaN     260.0  2008

### 2. 누락 데이터를 대체 값으로 채우기

대체 값으로 채우는 방법은 다양합니다.

통계값을 사용하거나, 이전 값/다음 값으로 채우기도 하고, 알고리즘이나 인공지능을 이용해 채우기도 합니다.

</br>

이번에는 mass 열의 NaN 데이터를 **평균값으로 대체**해볼 것입니다.

- Step 1) 먼저, mass 열의 평균값을 계산하고,
- Step 2) **`fillna`** 함수를 통해 NaN 데이터들을 평균값으로 대체해 보세요.

**주의)** **filled_df**를 바탕으로 진행하세요

**Q)** 결측치가 있는 열에 대해 mean을 해버리면 평균이 잘못 계산되지 않을까요?

**A)** mean 같은 통계 함수에는 자동으로 skipna 라는 매개변수가 True로 들어가게 됩니다.\
이는 결측치를 제외하고 계산하겠다는 의미입니다. 그렇기에 제대로 데이터에 대해서만 평균이 구해지게 됩니다.

In [ ]:
filled_df = dropped_df.copy()

# Step 1. 통계함수를 통해 mass 열의 평균값 계산
imputation = filled_df['mass'].mean(axis=0)

print(f"mass열의 평균치: {imputation}")

mass열의 평균치: 2.5093200401606426


In [ ]:
# Step 2. fillna 함수를 통해 distance 열의 NaN 데이터들을 평균값으로 대체
filled_df['mass'] = filled_df['mass'].fillna(imputation)

# inplace = True를 imputation 뒤에 넣을거면 변수로 따라 만들지 않고 하면됨. inplace는 새 데이터가 아닌 원본 자체를 바꾸는거임

# 누락 데이터 채운 후
filled_df.tail(5)

method  number  orbital_period     mass  distance  year
1030  Transit       1        3.941507  2.50932     172.0  2006
1031  Transit       1        2.615864  2.50932     148.0  2007
1032  Transit       1        3.191524  2.50932     174.0  2007
1033  Transit       1        4.125083  2.50932     293.0  2008
1034  Transit       1        4.187757  2.50932     260.0  2008

### 선택 사항)

- 다양한 통계함수를 사용해보세요.\
  통계함수 중에는 **평균값(mean), 중앙값(median), 최빈값(mode)**이 주로 사용됩니다.

---

## 범주형 데이터 처리
연속적인 실수 데이터가 아닌 경우 컴퓨터가 인식할 수 없기 때문에 이를 변환하는 과정이 필요합니다.\
그 중, 범주형 데이터는 대표적으로 0과 1로 표현되는 더미 변수를 사용합니다.

이번 실습에서는 데이터셋의 범주형 데이터를 더미 변수로 변환해보도록 하겠습니다.

In [ ]:
df.head(5)

method  number  orbital_period   mass  distance  year
0  Radial Velocity       1         269.300   7.10     77.40  2006
1  Radial Velocity       1         874.774   2.21     56.95  2008
2  Radial Velocity       1         763.000   2.60     19.84  2011
3  Radial Velocity       1         326.030  19.40    110.62  2007
4  Radial Velocity       1         516.220  10.50    119.47  2009

In [ ]:
print(df['method'].unique()) # 데이터의 종류를 알려주는 함수

['Radial Velocity' 'Imaging' 'Eclipse Timing Variations' 'Transit'
 'Astrometry' 'Transit Timing Variations' 'Orbital Brightness Modulation'
 'Microlensing' 'Pulsar Timing' 'Pulsation Timing Variations']


pandas의 get_dummies는 모든 범주형 변수를 더미 변수로 변환해주는 역할을 합니다.

특정 열 들만 더미 변수를 만들고 싶다면, columns 매개변수를 활용할 수 있습니다.

In [ ]:
df2 = pd.get_dummies(df, columns = ['method'])
# 판다스 함수이기 때문에 pd로 선언을 해주고 뒤에 df로 데이터 프레임 선언해주고 columns로 더미 데이터로 만들고 싶은 열을 넣으며 됨

df2.head(5)

number  orbital_period   mass  distance  year  method_Astrometry  \
0       1         269.300   7.10     77.40  2006              False   
1       1         874.774   2.21     56.95  2008              False   
2       1         763.000   2.60     19.84  2011              False   
3       1         326.030  19.40    110.62  2007              False   
4       1         516.220  10.50    119.47  2009              False   

   method_Eclipse Timing Variations  method_Imaging  method_Microlensing  \
0                             False           False                False   
1                             False           False                False   
2                             False           False                False   
3                             False           False                False   
4                             False           False                False   

   method_Orbital Brightness Modulation  method_Pulsar Timing  \
0                                 False                 False   
1                                 False                 False   
2                                 False                 False   
3                                 False                 False   
4                                 False                 False   

   method_Pulsation Timing Variations  method_Radial Velocity  method_Transit  \
0                               False                    True           False   
1                               False                    True           False   
2                               False                    True           False   
3                               False                    True           False   
4                               False                    True           False   

   method_Transit Timing Variations  
0                             False  
1                             False  
2                             False  
3                             False  
4                             False

---

## 정규화
변수가 갖는 범위의 크기는 변수마다 다를 수 있습니다.

이런 경우에는 데이터 분석할 때, 큰 범위를 갖는 변수가 더 큰 영향을 갖게 됩니다.

그래서 보통 모든 변수들이 같은 크기의 범위를 갖도록 **정규화**하는 작업을 가집니다.


In [ ]:
df = sns.load_dataset('planets')

df.dropna(subset=['mass', 'distance'], axis=0, inplace=True)

df.describe() # 데이터의 정보를 요약해서 보여주는 함수

number  orbital_period        mass    distance         year
count  498.00000      498.000000  498.000000  498.000000   498.000000
mean     1.73494      835.778671    2.509320   52.068213  2007.377510
std      1.17572     1469.128259    3.636274   46.596041     4.167284
min      1.00000        1.328300    0.003600    1.350000  1989.000000
25%      1.00000       38.272250    0.212500   24.497500  2005.000000
50%      1.00000      357.000000    1.245000   39.940000  2009.000000
75%      2.00000      999.600000    2.867500   59.332500  2011.000000
max      6.00000    17337.500000   25.000000  354.000000  2014.000000

위의 결과에서 세 가지의 연속형 변수를 살펴봅시다.

oribital_period는 범위가 `1.32` ~ `17337.5`인 것에 비해,\
mass는 `0` ~ `25`, distance는 `1.35` ~ `354`로 범위의 크기가 각자 다르다는 것을 알 수 있습니다.

데이터 분석을 위해선, 이 변수들의 범위를 모두 일정하게 맞춰야할 필요가 있습니다.

</br>


해당 실습에서는 distance 열에 대해서만 정규화를 진행해보도록 하겠습니다.

정규화 방법으로는 데이터의 범위를 0~1으로 변환하는 Min-Max Normalization을 사용해봅시다.

In [ ]:
#Step 1. 최솟값으로 빼서 가장 작은 값이 0이 되도록 변환
df['distance'] -= df['distance'].min()
# 예를 들어 3~ 10까지면 최솟값인 3을 빼주면 0~ 7까지로 되겠지

#Step 2. 데이터의 범위 크기를 측정
min_max = df['distance'].max() - df['distance'].min()
# 0~7이면 7 - 0을 해서 7이 되는거지

#Step 3. distance 열을 방금 구한 크기로 나눠 줌
df['distance'] = df['distance'] / min_max
# 0~ 7이니까 7로 나눠주면 0~1까지로 되겠지

print(df['distance'].describe())
# min과 max가 0과 1이어야 정답입니다.

count    498.000000
mean       0.143820
std        0.132131
min        0.000000
25%        0.065639
50%        0.109429
75%        0.164419
max        1.000000
Name: distance, dtype: float64


## 실습 종료 전 유의사항

※ 완료 시 우측 상단 프로필 옆 '**공유 - 액세스를 *링크가 있는 모든 사용자*로 전환**' 후 복사한 링크를 이루리 과제 란에 업로드 하세요.

<img src="https://drive.google.com/uc?id=1Zj1FKIz0sanqUfkfr3scp5M-dDgMFxYQ" height=300>

</br>

※ 추가로 상단 메뉴 '**파일 - 다운로드 - .ipynb**' 로 다운로드 후 첨부 파일로 업로드 하세요.

※ 즉, 과제의 텍스트란에는 링크를, 파일 첨부란에는 .ipynb 파일을 제출하십시오.